apis/projectcontour/v1alpha1/contourconfig.go

```diff
@@ -22,6 +25,396 @@ import (
 // It contains most of all the options that can be customized, the
 // other remaining options being command line flags.
 type ContourConfigurationSpec struct {
+	// XDSServer contains parameters for the xDS server.
+	// +optional
+	XDSServer ServerParameters `json:"xdsServer,omitempty"`
+
+	// IngressOptions contains parameters for ingress options.
+	// +optional
+	IngressOptions Ingress `json:"ingress,omitempty"`
+
+	EnvoyConfig Envoy `json:"envoy,omitempty"`
+
+	// GatewayConfig contains parameters for the gateway-api Gateway that Contour
+	// is configured to serve traffic.
+	// +optional
+	GatewayConfig *GatewayParameters `json:"gateway,omitempty"`
+
+	// HTTPProxy defines parameters on HTTPProxy.
+	// +optional
+	HTTPProxy *HTTPProxyConfig `json:"httpproxy,omitempty"`
+
+	// LeaderElection contains leader election parameters.
+	// +optional
+	LeaderElection LeaderElectionParameters `json:"leaderelection,omitempty"`
```

Comment:
Using this field as an example - currently, we have an [internal set of defaults](https://github.com/projectcontour/contour/blob/main/pkg/config/parameters.go#L722-L728) that are applied if the user does not specify values, but with the ConfigMap those defaults weren't really visible to the user anywhere other than documentation. Now that we're using a CRD, I'm thinking it would make sense to configure those as kubebuilder defaults, so that the default values are actually reflected in the CRD spec when a user creates a `ContourConfiguration`. WDYT? If that makes sense, then I guess we should go through all the defaults in the `Defaults()` func and set them up as kubebuilder defaults here.

Reply:
+1 for me, that sounds reasonable if it works well with kubebuilder etc.

Reply:
💯 percent agree that we should use CRD defaulting and move away from the `Defaults()` pattern wherever possible. We should then get documentation of the config format and its defaults for close to free with the already-existing reference generation that parses the types.

Reply:
Yup agree and doing that in many places. Just need to be aware of defaulting the wrong things (https://github.com/projectcontour/contour/pull/4013#discussion_r705359031)